tabpy-server/tabpy_server/app/app.py

```diff
@@ -0,0 +1,310 @@
+        self.subdirectory = ""
+        self.tabpy_state = None
+        self.python_service = None
+        self.credentials = {}
+
+        parser = configparser.ConfigParser()
+
+        if os.path.isfile(config_file):
+            with open(config_file) as f:
+                parser.read_string(f.read())
+        else:
+            logger.warning(
+                "Unable to find config file at '{}', "
+                "using default settings.".format(config_file))
+
+        def set_parameter(settings_key,
+                          config_key,
+                          default_val=None,
+                          check_env_var=False):
+            if config_key is not None and\
+               parser.has_option('TabPy', config_key):
+                self.settings[settings_key] = parser.get('TabPy', config_key)
+            elif check_env_var:
+                self.settings[settings_key] = os.getenv(
+                    config_key, default_val)
+            elif default_val is not None:
+                self.settings[settings_key] = default_val
+
+        set_parameter('port', ConfigParameters.TABPY_PORT,
+                      default_val=9004, check_env_var=True)
+        set_parameter('server_version', None, default_val=__version__)
+
+        set_parameter('upload_dir', ConfigParameters.TABPY_QUERY_OBJECT_PATH,
+                      default_val='/tmp/query_objects', check_env_var=True)
+        if not os.path.exists(self.settings['upload_dir']):
+            os.makedirs(self.settings['upload_dir'])
+
+        # set and validate transfer protocol
+        set_parameter('transfer_protocol',
+                      ConfigParameters.TABPY_TRANSFER_PROTOCOL,
+                      default_val='http')
+        self.settings['transfer_protocol'] =\
+            self.settings['transfer_protocol'].lower()
+
+        set_parameter('certificate_file',
+                      ConfigParameters.TABPY_CERTIFICATE_FILE)
+        set_parameter('key_file', ConfigParameters.TABPY_KEY_FILE)
+        self._validate_transfer_protocol_settings()
+
+        # if state.ini does not exist try and create it - remove
+        # last dependence on batch/shell script
+        set_parameter('state_file_path', ConfigParameters.TABPY_STATE_PATH,
+                      default_val='./', check_env_var=True)
+        self.settings['state_file_path'] = os.path.realpath(
+            os.path.normpath(
+                os.path.expanduser(self.settings['state_file_path'])))
+        state_file_path = self.settings['state_file_path']
+        logger.info("Loading state from state file %s" %
+                    os.path.join(state_file_path, "state.ini"))
+        tabpy_state = _get_state_from_file(state_file_path)
+        self.tabpy_state = TabPyState(
+            config=tabpy_state, settings=self.settings)
+
+        self.python_service = PythonServiceHandler(PythonService())
+        self.settings['compress_response'] = True if TORNADO_MAJOR >= 4\
```

Comment:
Is this needed since we require tornado 5.1.1 ?

Reply:
We require it for now but may relax requirements in the future.